ocaml/xapi/xapi_cluster_host.ml

```diff
@@ -31,110 +31,112 @@ let fix_pif_prerequisites ~__context (pif_ref,pif_rec) =
-    let body = Printf.sprintf "Unable to create cluster host on %s."
-        (Db.Host.get_name_label ~__context ~self:host) in
-    let obj_uuid = Db.Host.get_uuid ~__context ~self:host in
-    call_api_function_with_alert ~__context
-      ~msg:Api_messages.cluster_host_creation_failed
-      ~cls:`Host ~obj_uuid ~body
-      ~api_func:(fun rpc session_id -> ignore
-          (Client.Client.Cluster_host.create rpc session_id cluster_ref host pifref))
-  | None -> ()
+(* Create xapi db object for cluster_host, resync_host calls clusterd *)
+let create_internal ~__context ~cluster ~host ~pIF : API.ref_Cluster_host =
+  with_clustering_lock (fun () ->
+      assert_operation_host_target_is_localhost ~__context ~host;
+      let ref = Ref.make () in
+      let uuid = Uuidm.to_string (Uuidm.create `V4) in
+      Db.Cluster_host.create ~__context ~ref ~uuid ~cluster ~host ~pIF ~enabled:false
+        ~current_operations:[] ~allowed_operations:[] ~other_config:[] ~joined:false;
+      ref
+  )
 
-let resync_host ~__context ~host =
-  create_as_necessary ~__context ~host;
-  match (find_cluster_host ~__context ~host) with
-  | None              -> () (* no clusters exist *)
-  | Some cluster_host ->    (* cluster_host and cluster exist *)
-    (* Cluster_host.enable unconditionally invokes the low-level enable operations and is idempotent. *)
-    if Db.Cluster_host.get_enabled ~__context ~self:cluster_host then begin
-      (* RPU reformats partition, losing service status, never re-enables clusterd *)
-      debug "Cluster_host %s is enabled, starting up xapi-clusterd" (Ref.string_of cluster_host);
-      Xapi_clustering.Daemon.enable ~__context;
+(* Helper function atomically enables clusterd and joins the cluster_host *)
+let join_internal ~__context ~self =
+  with_clustering_lock (fun () ->
 
-      (* Best-effort Cluster_host.enable *)
-      let body = Printf.sprintf "Unable to create cluster host on %s."
-          (Db.Host.get_name_label ~__context ~self:host) in
-      let obj_uuid = Db.Host.get_uuid ~__context ~self:host in
-      call_api_function_with_alert ~__context
-        ~msg:Api_messages.cluster_host_enable_failed
-        ~cls:`Host ~obj_uuid ~body
-        ~api_func:(fun rpc session_id -> ignore
-            (Client.Client.Cluster_host.enable rpc session_id cluster_host))
-    end
+      let pIF = Db.Cluster_host.get_PIF ~__context ~self in
+      let pifrec = Db.PIF.get_record ~__context ~self:pIF in
+      fix_pif_prerequisites ~__context (pIF,pifrec);
 
-let create ~__context ~cluster ~host ~pif =
-  (* TODO: take network lock *)
-  with_clustering_lock (fun () ->
-      assert_operation_host_target_is_localhost ~__context ~host;
+      let host = Db.Cluster_host.get_host ~__context ~self in
       assert_cluster_host_can_be_created ~__context ~host;
-      assert_pif_attached_to ~host ~pif ~__context;
-      let ref = Ref.make () in
+      assert_pif_attached_to ~host ~pIF ~__context;
```

Comment:
We need to query the pif again, otherwise the fix PIF prerequisites function before has done its job of setting the PIF to disallow unplug, but we still got the old PIF.
Would be better if we moved querying the PIF inside fix_pif_prerequisites, and just give fix_pif_prerequisites the PIF ref, not the PIF record (which can be out of date)

Reply:
In that case, should all the PIF helpers use the ref and not the record?

Reply:
I think the code that modifies the PIF should be before the one that fetches the PIF record, so any code that modifies a PIF must not take the record as input, that way we can move the code that fetches the PIF record after all the code that modifies it. Currently the only code that modifies it is `fix_pif_prerequisites`